File: packages/scheduling-engine/src/scheduler/block-factories.ts
// =====================================================================
//  INCOMPOL PLAN -- Block Factories
//  Factory functions for creating blocked, overflow, and infeasible blocks.
//  Extracted from slot-allocator.ts
// =====================================================================

import { S0 } from '../constants.js';
import type { Block } from '../types/blocks.js';
import type { InfeasibilityReason } from '../types/infeasibility.js';
import type { SkuBucket, ToolGroup } from './demand-grouper.js';

export function mkBlocked(sk: SkuBucket, grp: ToolGroup, di: number, reason: string): Block {
  return {
    opId: sk.opId,
    toolId: grp.toolId,
    sku: sk.sku,
    nm: sk.nm,
    machineId: grp.machineId,
    origM: sk.origM,
    dayIdx: di,
    eddDay: sk.edd,
    qty: 0,
    prodMin: 0,
    setupMin: 0,
    operators: sk.operators,
    blocked: true,
    reason,
    moved: sk.moved,
    hasAlt: sk.hasAlt,
    altM: sk.altM,
    mp: sk.mp,
    stk: sk.stk,
    lt: sk.lt,
    atr: sk.atr,
    startMin: S0,
    endMin: S0,
    setupS: null,
    setupE: null,
    type: 'blocked',
    shift: 'X',
  };
}

export function mkOverflow(sk: SkuBucket, grp: ToolGroup, di: number, ofMin: number): Block {
  return {
    opId: sk.opId,
    toolId: grp.toolId,
    sku: sk.sku,
    nm: sk.nm,
    machineId: grp.machineId,
    origM: sk.origM,
    dayIdx: di,
    eddDay: sk.edd,
    qty: 0,
    prodMin: sk.prodMin,
    setupMin: 0,
    operators: sk.operators,
    blocked: false,
    reason: null,
    moved: sk.moved,
    hasAlt: sk.hasAlt,
    altM: sk.altM,
    mp: sk.mp,
    stk: sk.stk,
    lt: sk.lt,
    atr: sk.atr,
    startMin: S0,
    endMin: S0,
    setupS: null,
    setupE: null,
    type: 'overflow',
    shift: 'X',
    overflow: true,
    overflowMin: ofMin,
  };
}

export function mkInfeasible(
  sk: SkuBucket,
  grp: ToolGroup,
  di: number,
  reason: InfeasibilityReason,
  detail: string,
): Block {
  return {
    opId: sk.opId,
    toolId: grp.toolId,
    sku: sk.sku,
    nm: sk.nm,
    machineId: grp.machineId,
    origM: sk.origM,
    dayIdx: di,
    eddDay: sk.edd,
    qty: 0,
    prodMin: sk.prodMin,
    setupMin: 0,
    operators: sk.operators,
    blocked: false,
    reason: null,
    moved: sk.moved,
    hasAlt: sk.hasAlt,
    altM: sk.altM,
    mp: sk.mp,
    stk: sk.stk,
    lt: sk.lt,
    atr: sk.atr,
    startMin: S0,
    endMin: S0,
    setupS: null,
    setupE: null,
    type: 'infeasible',
    shift: 'X',
    infeasibilityReason: reason,
    infeasibilityDetail: detail,
  };
}

File: apps/backend/src/domain/scheduling/scheduler/block_factories.py
"""Block factory functions — port of scheduler/block-factories.ts.

Creates blocked, overflow, and infeasible Block instances.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..constants import S0
from ..types import Block

if TYPE_CHECKING:
    from .demand_grouper import SkuBucket, ToolGroup


def mk_blocked(sk: SkuBucket, grp: ToolGroup, di: int, reason: str) -> Block:
    return Block(
        op_id=sk["op_id"],
        tool_id=grp["tool_id"],
        sku=sk["sku"],
        nm=sk["nm"],
        machine_id=grp["machine_id"],
        orig_m=sk["orig_m"],
        day_idx=di,
        edd_day=sk["edd"],
        qty=0,
        prod_min=0,
        setup_min=0,
        operators=sk["operators"],
        blocked=True,
        reason=reason,
        moved=sk["moved"],
        has_alt=sk["has_alt"],
        alt_m=sk.get("alt_m"),
        mp=sk.get("mp"),
        stk=sk.get("stk", 0),
        lt=sk.get("lt", 0),
        atr=sk.get("atr", 0),
        start_min=S0,
        end_min=S0,
        type="blocked",
        shift="X",
    )


def mk_overflow(sk: SkuBucket, grp: ToolGroup, di: int, of_min: float) -> Block:
    return Block(
        op_id=sk["op_id"],
        tool_id=grp["tool_id"],
        sku=sk["sku"],
        nm=sk["nm"],
        machine_id=grp["machine_id"],
        orig_m=sk["orig_m"],
        day_idx=di,
        edd_day=sk["edd"],
        qty=0,
        prod_min=int(sk["prod_min"]),
        setup_min=0,
        operators=sk["operators"],
        blocked=False,
        moved=sk["moved"],
        has_alt=sk["has_alt"],
        alt_m=sk.get("alt_m"),
        mp=sk.get("mp"),
        stk=sk.get("stk", 0),
        lt=sk.get("lt", 0),
        atr=sk.get("atr", 0),
        start_min=S0,
        end_min=S0,
        type="overflow",
        shift="X",
        overflow=True,
        overflow_min=int(of_min),
    )


def mk_infeasible(
    sk: SkuBucket,
    grp: ToolGroup,
    di: int,
    reason: str,
    detail: str,
) -> Block:
    return Block(
        op_id=sk["op_id"],
        tool_id=grp["tool_id"],
        sku=sk["sku"],
        nm=sk["nm"],
        machine_id=grp["machine_id"],
        orig_m=sk["orig_m"],
        day_idx=di,
        edd_day=sk["edd"],
        qty=0,
        prod_min=int(sk["prod_min"]),
        setup_min=0,
        operators=sk["operators"],
        blocked=False,
        moved=sk["moved"],
        has_alt=sk["has_alt"],
        alt_m=sk.get("alt_m"),
        mp=sk.get("mp"),
        stk=sk.get("stk", 0),
        lt=sk.get("lt", 0),
        atr=sk.get("atr", 0),
        start_min=S0,
        end_min=S0,
        type="infeasible",
        shift="X",
        infeasibility_reason=reason,
        infeasibility_detail=detail,
    )
